core/Cookie.php

```diff
@@ -243,13 +253,18 @@ protected function loadContentFromCookie($validateSignature = true)
             $varName = substr($nameValue, 0, $equalPos);
             $varValue = substr($nameValue, $equalPos + 1);
 
-            // no numeric value are base64 encoded so we need to decode them
             if (!is_numeric($varValue)) {
                 $tmpValue = base64_decode($varValue);
-                $varValue = safe_unserialize($tmpValue);
+                if ($isSigned) {
+                    // only unserialise content if it was signed meaning the cookie was generated pre Matomo 4
+                    $varValue = safe_unserialize($tmpValue);
```

Comment:
`safe_unserialize` uses our custom unserialize code which doesn't use `unserialize` at all, so I think this is fine? I guess it depends on how the cookie values are used to say if it's completely safe from injection. I guess we could also make sure there are only strings & numbers in the unserialized array to make it a little safer?

Reply:
AFAIK the most risk re safe_unserialize would be from when PHP actually does the unserialize. This risk should be eliminated with the `safe_unserialize` (side note: I think since we now require PHP 7.2+ we could even use PHP unserialize with the additional parameter).

Not sure there's much benefit in checking it additionally. I guess with this PR people would need to know the general salt in order for this to be used so it shouldn't be much of an issue anymore anyway since newly signed cookies would no longer be signed

Reply:
safe_unserialize is defined in /libs/upgradephp/upgrade.php (included in bootstrap.php), and it's our code. I don't think it's PHP API? So should be super safe, I think. Unless we, use a variable in an insecure way, like modifying a session variable based on the index we get from the cookie or something like that. This is what I meant at least w/ my comment.

Reply:
Sorry I meant eventually we can use https://www.php.net/manual/en/function.unserialize.php like `unserialize ($str, false)` which I think should be the same as our safe_unserialize I reckon (but didn't fully check). PHP5.X didn't support the `allowed_classes` parameter

Reply:
Imho it's fine to simply use `safe_unserialize` everywhere. It automatically disallows all classes, if non where given. Also it catches all errors and logs them. Should be simpler than doing that manually everywhere again...

Reply:
Just keep in mind that the `safe_unserialize` in Matomo is over 10 years old and might have some weird bugs in it. (The upgrade.php seems to be only intended for `PHP_VERSION<5.2`.

But it should be pretty safe as it only uses regex (unless someone finds an input value that creates a DOS).

Reply:
Oh. sure. Thought that was the method from Common class 🙈 We shouldn't use the `safe_unserialize` from `upgrade.php`. Should be `Common::safe_unserialize` for sure!

Reply:
Hm, now that only PHP7 is supported we might be able to remove that upgrade.php completely. I'll see what happens when I try.

Reply:
I thought initially as well it could be removed but reading https://www.php.net/manual/en/function.unserialize.php and the warning it says clearly regardless of any options value it isn't safe for user input (which it may be here) so prefer to maybe keep the existing safe_unserialise usage here for now. It should be mostly only for previously created cookies anyway. Unless someone had the general salt then they could still create any user content for cookie and sign it and it would use unserialize.